Ad9545 zero delay mode

tachicialex opened the pull request:
e32665df9179 - clk: ad9545: Fix free run rate recalc
Rate recalculation for the free run mode was wrong.

38eb3f601d05 - clk: ad9545: Fix validation timer write
Validation timer would be read from DT but not programmed.

9476bfe5ed52 - clk: ad9545: fix output current selection
Current selection was using the wrong iterator.

60229f6746b6 - iio: hmc7044: add n-shot trigger
Allow user to trigger a sync pulse from a clock mentioned in DT.

19cb9b4bb19d - clk: ad9545: add zero delay support
Mode in which AD9545 aligns the output phase to the input phase.

c7858a6aa38b - blindings: clock: ad9545: document use of slew rate
Allow the user to program the slew rate from DT. Application depended as some ICs are sensible to high output slew-rates.

6994adbf4c02 - clk: clk-conf: allow user to setup an n-shot from DT
e89531db498c - bindings: clock: ad9545: document use of zero delay

nunojsa reviewed (commented):
Overall looks good. Some concerns regarding dts and changes made in the core... Regarding the fixes, should we backport them to 2019_R2? Also, the fixes should be the first patches in the series so that they don't depend on other patches and it's easier to backport them

tachicialex commented on Documentation/devicetree/bindings/clock/clk-ad9545.yaml:
```diff
@@ -267,6 +267,22 @@ patternProperties:
+          rate to become imutable.
+        $ref: /schemas/types.yaml#/definitions/uint32
+        enum: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
```
No, device won't need any special wiring. Added these here as  a way to switch between zero-delay and Phase Build-Out modes. (zero delay being less flexible but provides  a deterministic output phase)

nunojsa replied:
Hmm, than this might be questionable... Anyways, I'm also not sure how to add these kind of setting on clock providers. So this might be acceptable upstream

tachicialex commented on drivers/clk/clk-conf.c:
```diff
@@ -139,6 +183,10 @@ int of_clk_set_defaults(struct device_node *node, bool clk_supplier)
+	if (rc < 0)
+		return rc;
+
+	return __set_clk_nshot(node, clk_supplier);
```
Yes, after ad9545 gets through, ill send this also.

tachicialex commented on Documentation/devicetree/bindings/clock/clk-ad9545.yaml:
```diff
@@ -283,6 +283,13 @@ patternProperties:
+          DPLL0 phase slew limit rate (in picoseconds/second).
+        $ref: /schemas/types.yaml#/definitions/uint32
+        minimum: 1
+        maximum: 4000000000
```
Talked to the chip designer. Slew rate needs to be limited if the IC connected to the AD9545's output cannot handle quick phase adjustments.

tachicialex commented:
Changelog V1 -> V2:
	- removed -hz from adi,pll-internal-zero-delay-hz
	- in clk-conf, if unable to set nshot, print warn, not err
	- called of_node_put on error paths and end of function
	- added checks for parsed DT values for: pll-internal-zero-delay-feedback path and rate.
	- setup the n-shot before enabling the clock (before this, hmc7044 assumed sync_clk was
	already configured to output a 1 pulse shot at enable)

nunojsa commented on drivers/clk/clk-conf.c:
```diff
@@ -116,6 +116,50 @@ static int __set_clk_rates(struct device_node *node, bool clk_supplier)
+				       __clk_get_name(clk), nshot, rc,
+				       clk_get_nshot(clk));
+
+			clk_put(clk);
```
I still think `clk_put()` has to be called in error paths...

nunojsa commented on drivers/iio/frequency/hmc7044.c:
```diff
@@ -607,6 +608,30 @@ static ssize_t hmc7044_sync_pin_mode_store(struct device *dev,
+		 */
+		i = clk_get_nshot(sync_clk);
+		if (i < 0)
+			return ret;
```
no `clk_put()`??

tachicialex replied:
In hmc7044, i solved it. Not sure where do you want me to add clk_put in [drivers/clk/clk-conf.c](https://github.com/analogdevicesinc/linux/pull/1792/files/e89531db498c78ce5b3781fe203c979a72db08ba..6994adbf4c02be9090421dffa61a49ee39c93656#diff-5f59bde8203b44f79bbcd8386b38443a2f44b8c15b8d5f1ef84fa5879798edb3) ?

nunojsa commented on drivers/iio/frequency/hmc7044.c:
```diff
@@ -607,6 +608,30 @@ static ssize_t hmc7044_sync_pin_mode_store(struct device *dev,
+		}
+
+		clk_prepare_enable(sync_clk);
+		clk_disable_unprepare(sync_clk);
```
Why toggling between enable/disable? Again, a comment is definitely helpful for others that read this code

nunojsa reviewed (commented):
Please also address the checkpatch warning...

tachicialex commented:
Changelog V2 -> V3:
- added clk_put on error path in hmc7044
- fixed checkpatch warnings

nunojsa commented:
Something went wrong with the checkpatch job. You might have to rebase your branch on master...

dbogdan commented:
> Something went wrong with the checkpatch job. You might have to rebase your branch on master...

V4:
 - use the current master branch as base for these changes.
